engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,671 @@
+            downstream.modifiedColumnSet = result.getModifiedColumnSetForUpdates();
+            downstream.shifted = RowSetShiftData.EMPTY;
+
+            boolean resultModified = false;
+
+            for (MultiJoinListenerRecorder recorder : recorders) {
+                if (recorder.recordedVariablesAreValid()) {
+                    if (recorder.getRemoved().isNonempty() || recorder.getAdded().isNonempty()) {
+                        if (recorder.getParent().size() > 1) {
+                            throw new IllegalStateException(
+                                    "Multiple rows in " + recorder.tableNumber + " for zero-key multiJoin.");
+                        }
+                        resultModified = true;
+                        redirections[recorder.tableNumber].writableSingleValueCast()
+                                .setValue(recorder.getParent().getRowSet().firstRowKey());
+                        downstream.modifiedColumnSet.setAll(modifiedColumnSets[recorder.tableNumber]);
+                    } else if (recorder.getModified().isNonempty()) {
+                        resultModified = true;
+                        recorder.transformer.transform(recorder.getModifiedColumnSet(), downstream.modifiedColumnSet);
+                        redirections[recorder.tableNumber].writableSingleValueCast()
+                                .setValue(recorder.getParent().getRowSet().firstRowKey());
+                    } else if (recorder.getShifted().nonempty()) {
+                        redirections[recorder.tableNumber].writableSingleValueCast()
+                                .setValue(recorder.getParent().getRowSet().firstRowKey());
+                    }
+                }
+            }
+
+            final boolean hasResults = Arrays.stream(redirections).anyMatch(rd -> rd.getValue() != NULL_ROW_KEY);
+            if (hasResults && result.size() == 0) {
+                result.getRowSet().writableCast().insert(0);
+                downstream.added = RowSetFactory.fromKeys(0);
+                downstream.removed = RowSetFactory.empty();
+                downstream.modified = RowSetFactory.empty();
+                downstream.modifiedColumnSet = ModifiedColumnSet.EMPTY;
+            } else if (!hasResults && result.size() == 1) {
+                result.getRowSet().writableCast().remove(0);
+                downstream.added = RowSetFactory.empty();
+                downstream.removed = RowSetFactory.fromKeys(0);
+                downstream.modified = RowSetFactory.empty();
+                downstream.modifiedColumnSet = ModifiedColumnSet.EMPTY;
+            } else {
+                downstream.added = RowSetFactory.empty();
+                downstream.removed = RowSetFactory.empty();
+                if (resultModified) {
+                    downstream.modified = RowSetFactory.fromKeys(0);
+                } else {
+                    // this would be a useless update
+                    return;
```

Comment:
I think we don't need to close the update, since we never set any fields on it, right?